### PR TITLE
internal: add json `tracing` Layer for profiling startup

### PR DESCRIPTION
**Pull request by davidbarsky**
On `buck2/integrations/rust-project`, this results in the following being printed:

```json
{"name":"discover_command","elapsed_ms":18703}
{"name":"parallel_prime_caches","elapsed_ms":0}
{"name":"vfs_load","elapsed_ms":5895}
{"name":"vfs_load","elapsed_ms":547}
{"name":"parallel_prime_caches","elapsed_ms":23}
{"name":"parallel_prime_caches","elapsed_ms":84}
{"name":"parallel_prime_caches","elapsed_ms":5819}
```

### CODE REVIEW

File: crates/rust-analyzer/src/tracing/config.rs [hunk 1]
```diff
@@ -51,31 +53,48 @@ where
...
-        let profiler_layer = self.profile_filter.map(|spec| hprof::layer(&spec));
+        let profiler_layer = match self.profile_filter {
+            Some(spec) => hprof::layer(&spec).boxed(),
+            None => hprof::layer("").with_filter(LevelFilter::OFF).boxed(),
+        };
+
+        let json_profiling_filter = filter_fn(|metadata| {
+            let name = metadata.name();
+            let matches_name =
+                name == "vfs_load" || name == "parallel_prime_caches" || name == "discover_command";
```
Comment (ChayimFriedman2) [inline crates/rust-analyzer/src/tracing/config.rs:86]: Why are you limiting to those spans? It seems more general to allow any span with filtering set by the user (`RA_PROFILE`).

**Comment by davidbarsky** [comment]
The following—which is in this PR—works properly and doesn't result in [this event](https://github.com/rust-lang/rust-analyzer/blob/28830ff2f1158ee92f4852ef3ec35af0935d1562/crates/hir-ty/src/chalk_db.rs#L849-L855) being logged:

```rust
impl<T> Config<T>
where
    T: for<'writer> MakeWriter<'writer> + Send + Sync + 'static,
{
    pub fn init(self) -> anyhow::Result<()> {
        let targets_filter: Targets = self
            .filter
            .parse()
            .with_context(|| format!("invalid log filter: `{}`", self.filter))?;

        let writer = self.writer;

        let ra_fmt_layer = tracing_subscriber::fmt::layer()
            .with_target(false)
            .with_ansi(false)
            .with_writer(writer);

        let chalk_layer = match self.chalk_filter {
            Some(chalk_filter) => {
                let level: LevelFilter =
                    chalk_filter.parse().with_context(|| "invalid chalk log filter")?;

                let chalk_filter = Targets::new()
                    .with_target("chalk_solve", level)
                    .with_target("chalk_ir", level)
                    .with_target("chalk_recursive", level);
                HierarchicalLayer::default()
                    .with_indent_lines(true)
                    .with_ansi(false)
                    .with_indent_amount(2)
                    .with_writer(io::stderr)
                    .with_filter(chalk_filter)
                    .boxed()
            }
            None => HierarchicalLayer::default().with_filter(LevelFilter::OFF).boxed(),
        };

        let profiler_layer = match self.profile_filter {
            Some(spec) => hprof::layer(&spec).boxed(),
            None => hprof::layer("").with_filter(LevelFilter::OFF).boxed(),
        };

        let json_profiling_filter = filter_fn(|metadata| {
            let name = metadata.name();
            let matches_name =
                name == "vfs_load" || name == "parallel_prime_caches" || name == "discover_command";

            matches_name && metadata.is_span()
        });

        let subscriber = Registry::default()
            .with(ra_fmt_layer.with_filter(targets_filter))
            .with(hprof::JsonTimingLayer.with_filter(json_profiling_filter))
            .with(profiler_layer)
            .with(chalk_layer);

        tracing::subscriber::set_global_default(subscriber)?;

        Ok(())
    }
}
```

The following, however, *does* result in [this event](https://github.com/rust-lang/rust-analyzer/blob/28830ff2f1158ee92f4852ef3ec35af0935d1562/crates/hir-ty/src/chalk_db.rs#L849-L855) being logged:

```rust
impl<T> Config<T>
where
    T: for<'writer> MakeWriter<'writer> + Send + Sync + 'static,
{
    pub fn init(self) -> anyhow::Result<()> {
        let targets_filter: Targets = self
            .filter
            .parse()
            .with_context(|| format!("invalid log filter: `{}`", self.filter))?;

        let writer = self.writer;

        let ra_fmt_layer = tracing_subscriber::fmt::layer()
            .with_target(false)
            .with_ansi(false)
            .with_writer(writer);

        let mut chalk_layer = None;
        if let Some(chalk_filter) = self.chalk_filter {
            let level: LevelFilter =
                chalk_filter.parse().with_context(|| "invalid chalk log filter")?;

            let chalk_filter = Targets::new()
                .with_target("chalk_solve", level)
                .with_target("chalk_ir", level)
                .with_target("chalk_recursive", level);
            chalk_layer = Some(
                HierarchicalLayer::default()
                    .with_indent_lines(true)
                    .with_ansi(false)
                    .with_indent_amount(2)
                    .with_writer(io::stderr)
                    .with_filter(chalk_filter),
            );
        };

        let mut profiler_layer = None;
        if let Some(spec) = self.profile_filter {
            profiler_layer = Some(hprof::layer(&spec));
        }

        let json_profiling_filter = filter_fn(|metadata| {
            let name = metadata.name();
            let matches_name =
                name == "vfs_load" || name == "parallel_prime_caches" || name == "discover_command";

            matches_name && metadata.is_span()
        });

        let subscriber = Registry::default()
            .with(ra_fmt_layer.with_filter(targets_filter))
            .with(hprof::JsonTimingLayer.with_filter(json_profiling_filter))
            .with(profiler_layer)
            .with(chalk_layer);

        tracing::subscriber::set_global_default(subscriber)?;

        Ok(())
    }
}
```

I'll try and get a minimal reproduction on tracing-subscriber.

File: crates/rust-analyzer/src/tracing/config.rs (hunk 1 above)
Comment (davidbarsky) [reply crates/rust-analyzer/src/tracing/config.rs:86]: I was mostly interested in measuring startup, but I'll add another env variable for (`RA_PROFILE_JSON`?) to make this customizable.

(I meant to open this as a draft)

Comment (ChayimFriedman2) [reply crates/rust-analyzer/src/tracing/config.rs:86]: I think it is fine to reuse `RA_PROFILE` for that.

Comment (davidbarsky) [reply crates/rust-analyzer/src/tracing/config.rs:86]: I respectfully disagree: I intend for the JSON output to be consumed by an LSP proxy that'll be taking these timings and converting them into metrics, and I don't really want to deal with the regular `RA_PROFILE` hprof output on the consumption side of things. I think `RA_PROFILE_JSON` should be used by machines, whereas I think `RA_PROFILE` should be used by people. I generally expect these to be mutually exclusive, tbh.

Comment (davidbarsky) [reply crates/rust-analyzer/src/tracing/config.rs]: I'll mark this as resolved, as I've added `RA_PROFILE_JSON`.

**Pull request merged**